x-pack/plugins/ingest_manager/server/services/epm/fields/field.ts

```diff
@@ -43,45 +49,133 @@ export interface Field {
 export type Fields = Field[];
 
 /**
- * ProcessFields takes the given fields read from yaml and expands it.
+ * expandFields takes the given fields read from yaml and expands them.
  * There are dotted fields in the field.yml like `foo.bar`. These should
- * be stored as an object inside an object and is the main purpose of this
- * preprocessing.
+ * be stored as an field within a 'group' field.
  *
- * Note: This function modifies the passed field param.
+ * Note: This function modifies the passed fields array.
  */
-export function processFields(fields: Fields) {
+export function expandFields(fields: Fields) {
   fields.forEach((field, key) => {
     const fieldName = field.name;
 
-    // If the field name contains a dot, it means we need to create sub objects
+    // If the field name contains a dot, it means we need to
+    // - take the first part of the name
+    // - create a field of type 'group' with this first part
+    // - put the original field, named with the rest of the original name in the fields property of the new group field
     if (fieldName.includes('.')) {
       // Split up the name by dots to extract first and other parts
       const nameParts = fieldName.split('.');
 
       // Getting first part of the name for the new field
-      const newNameTop = nameParts[0];
-      delete nameParts[0];
+      const groupFieldName = nameParts[0];
 
       // Put back together the parts again for the new field name
-      const newName = nameParts.length === 1 ? nameParts[0] : nameParts.slice(1).join('.');
+      const restFieldName = nameParts.slice(1).join('.');
 
-      field.name = newName;
+      // keep all properties of the original field, but give it the shortened name
+      field.name = restFieldName;
 
-      // Create the new field with the old field inside
-      const newField: Field = {
-        name: newNameTop,
+      // create a new field of type group with the original field in the fields array
+      const groupField: Field = {
+        name: groupFieldName,
         type: 'group',
         fields: [field],
       };
-      // Replace the old field in the array
-      fields[key] = newField;
-      if (newField.fields) {
-        processFields(newField.fields);
+      // check child fields further down the tree
+      if (groupField.fields) {
+        expandFields(groupField.fields);
+      }
+      // Replace the original field in the array with the new one
+      fields[key] = groupField;
+    } else {
+      // even if this field doesn't have dots to expand, its child fields further down the tree might
+      if (field.fields) {
+        expandFields(field.fields);
       }
     }
   });
 }
+/**
+ * dedupFields takes the given fields and merges sibling fields with the
+ * same name together.
+ * These can result from expandFields when the input contains dotted field
+ * names that share parts of their hierarchy.
+ */
+function dedupFields(fields: Fields): Fields {
```

Comment:
if there are to fields that have the same name but  are not a group type, it throws an error.  Is that desirable?  I thought this was possible.

Reply:
The two fields could be acceptable if the are exactly the same. I can amend that.

If they differ in any way (different type, or for some types, different in other properties) that would result in two different mappings for the same field present. Even if ES accepts that, I don't think it is desirable.

Reply:
I guess what I mean is, I thought its possible that we have duplicate fields we want to dedupe, but not necessarily throw an error?  This might explain the error I get when I try to install nginx?

`
UnhandledPromiseRejectionWarning: Error: Can't merge fields {"name":"answers","level":"extended","type":"object","object_type":"keyword","description":"An array containing an object for each answer section returned by the server.\nThe main keys that should be present in these objects are defined by ECS. Records that have more information may contain more keys than what ECS defines.\nNot all DNS data sources give all details about DNS answers. At minimum, answer objects must contain the `data` key. If more information is available, map as much of it to ECS as possible, and add any additional fields to the answer objects as custom fields."} and {"name":"answers","type":"group","fields":[{"name":"class","level":"extended","type":"keyword","ignore_above":1024,"description":"The class of DNS data contained in this resource record.","example":"IN"}]}
`

Reply:
Okay. If two fields are exactly the same and both are group types with fields, I don't think it should throw an error, but continue. If they have the same name, but different types, we should perhaps continue as well?  I'm not sure if we should throw an error and stop the install.   @ruflin what do you think?

Reply:
It depends. If we create the Kibana index pattern, I think we should ignore some errors as the data might come from different packages and there might be some conflicts (there shouldn't be). If we generate the template for a package, all should fit perfectly. In the best case we detect it already when building the package that something is off.

What should be possible is that the same field shows up twice and one has more details. For example `foo` is `keyword` and later also as `ignore_above: 1024`. But in general we should try to keep things as simple as possible as this kind of inheritance is a problem for index templates today.

My suggestion: Lets ignore errors for now as part of this PR to keep it moving forward, we need all the other changes in this PR. And then we follow up on a discussion on how we handle conflicts exactly.